### ISSUE TITLE
Enable setting AWS API Gateway integration timeout

### ISSUE THREAD

**Issue by danmactough** [serverless/serverless#12800]
### Is there an existing issue for this?

- [X] I have searched existing issues, it hasn't been reported yet

### Use case description

Currently, there is no way to configure the [integration timeout](https://docs.aws.amazon.com/apigateway/latest/developerguide/limits.html#api-gateway-execution-service-limits-table), and when [generating the method integration (for a REST API)](https://github.com/serverless/serverless/blob/de62c71e30855eff688f032ff10b9ad22de13afc/lib/plugins/aws/package/compile/events/api-gateway/lib/method/integration.js#L52), the [optional `TimeoutInMillis`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-apigateway-method-integration.html#cfn-apigateway-method-integration-timeoutinmillis) property is omitted (thus defaulting to 29000ms).

While this has previously prevent folks from adjusting the integration timeout to a shorter value, it has not really been a problem. But [AWS recently removed the hard-cap of 29s for REST APIs](https://aws.amazon.com/about-aws/whats-new/2024/06/amazon-api-gateway-integration-timeout-limit-29-seconds/). Folks who want to increase the timeout can [submit a request to increase the service quota](https://us-east-1.console.aws.amazon.com/servicequotas/home/services/apigateway/quotas/L-E5AE38E3) (my request was automatically approved).

And while users who have increased the API Gateway timeout [can configure their Lambdas to timeout after more than 29s](https://www.serverless.com/framework/docs/providers/aws/guide/serverless.yml#general-aws-lambda-settings) (notwithstanding the outdated warning about the API Gateway maximum timeout), we cannot adjust the API Gateway integration timeout.

Serverless should provide a way for users to update the integration timeout for REST APIs.

FWIW, my team is currently working around this issue using the following custom plugin to monkey-patch the `getMethodIntegration` function:

```js
const assert = require('node:assert');
const Klass = require('serverless/lib/plugins/aws/package/compile/events/api-gateway');

class IntegrationTimeout {
  constructor(serverless, options) {
    this.serverless = serverless;
    this.options = options;
    const instance = serverless.pluginManager.plugins.find(
      (plugin) => plugin.constructor.name === Klass.name
    );
    if (serverless.service.provider.timeout) {
      assert.ok(instance, `Could not find instance of ${Klass.name}`);
      const original = instance.getMethodIntegration;
      instance.getMethodIntegration = function (
        http,
        { lambdaLogicalId, lambdaAliasName }
      ) {
        const integration = original.call(instance, http, {
          lambdaLogicalId,
          lambdaAliasName,
        });
        const type = http.integration || 'AWS_PROXY';
        if (type === 'AWS' || type === 'AWS_PROXY') {
          integration.Properties.Integration.TimeoutInMillis =
            serverless.service.provider.timeout * 1000;
        }
        return integration;
      };
    }
  }
}

module.exports = IntegrationTimeout;
```

## Notes

- We are re-using `provider.timeout` for the value of the integration timeout. That works for us, but may not be the right decision for Serverless.
- We haven't done anything about the warning: "Warning: Function [function name] has timeout of ____ seconds, however, it's attached to API Gateway so it's automatically limited to 30 seconds." I think Serverless will want to resolve that warning.

**Comment by johan1252** [comment]
Thanks for the workaround.

Ended up doing very similar thing but added a `timeout` field under `events.http` so it could be different (or not defined) for individual endpoints.
Example:
1. Plugin - `updateAPIGatewayIntegrationTimeout.js`
    ```
    const assert = require("node:assert");
    const Klass = require("serverless/lib/plugins/aws/package/compile/events/api-gateway");
    
    class IntegrationTimeout {
	    constructor(serverless, options) {
		    serverless.configSchemaHandler.defineFunctionEventProperties("aws", "http", {
			    properties: {
				    timeout: { type: "number" }
			    }
		    });
    
		    this.serverless = serverless;
		    this.options = options;
		    const instance = serverless.pluginManager.plugins.find(
			    plugin => plugin.constructor.name === Klass.name
		    );
		    assert.ok(instance, `Could not find instance of ${Klass.name}`);
		    const original = instance.getMethodIntegration;
		    instance.getMethodIntegration = function (http, { lambdaLogicalId, lambdaAliasName }) {
			    const integration = original.call(instance, http, {
				    lambdaLogicalId,
				    lambdaAliasName
			    });
			    const type = http.integration || "AWS_PROXY";
			    if (http.timeout) {
				    if (type === "AWS" || type === "LAMBDA" || type === "AWS_PROXY") {
					    integration.Properties.Integration.TimeoutInMillis = http.timeout * 1000;
				    }
			    }
			    return integration;
		    };
	    }
    }
    
    module.exports = IntegrationTimeout;
    ```
2. Import plugin - `serverless.yml`
    ```
    plugins:
      - ./updateAPIGatewayIntegrationTimeout
    ```
3. Add timeout to endpoints that need it
    ```
    myAPI:
        handler: src/handlers/myAPI.handler
        timeout: 60 # Lambda timeout
        events:
          - http:
              authorizer: aws_iam
              integration: lambda
              path: my-api
              method: get
              timeout: 60 # New API Gateway timeout field
    ```

**Comment by nicksmithr** [comment]
Hey folks, would also love to get this functionality added.
I tried the above workaround on v4.4.0 but I think the serverless package is no longer bundled in the same way.

**Comment by danmactough** [comment]
> I tried the above workaround on v4.4.0 but I think the serverless package is no longer bundled in the same way.

Yeah, this is my fear with this monkey-patch.

I'd be happy to pick up this feature and open a PR with a proper implementation if the project leaders want it.

**Comment by ashishjullia** [comment]
@danmactough Is there a way to do this via embedding cloudformation inside serverless.yml? will it be a good idea and safe in same context?

**Comment by czubocha** [comment]
Thank you all for your input! The ability to set the AWS API Gateway integration timeout has been implemented and is now available in [v4.4.13](https://github.com/serverless/serverless/releases/tag/v4.4.13). You can find the docs [here](https://www.serverless.com/framework/docs/providers/aws/events/apigateway#timeout).
Feel free to share any further issues, feature requests, or feedback!
